ai_edge_torch/generative/examples/stable_diffusion/pipeline.py

```diff
@@ -0,0 +1,189 @@
+    uncond_prompt = ''
+
+  if sampler == 'k_lms':
+    sampler = samplers.KLMSSampler(n_inference_steps=n_inference_steps)
+  elif sampler == 'k_euler':
+    sampler = samplers.KEulerSampler(n_inference_steps=n_inference_steps)
+  elif sampler == 'k_euler_ancestral':
+    sampler = samplers.KEulerAncestralSampler(n_inference_steps=n_inference_steps)
+  else:
+    raise ValueError(
+        'Unknown sampler value %s. '
+        'Accepted values are {k_lms, k_euler, k_euler_ancestral}' % sampler
+    )
+
+  # Text embedding.
+  cond_tokens = model.tokenizer.encode(prompt)
+  cond_context = model.clip(np.array(cond_tokens), signature_name='encode')
+  uncond_tokens = model.tokenizer.encode(uncond_prompt)
+  uncond_context = model.clip(np.array(uncond_tokens), signature_name='encode')
+  context = np.concatenate([cond_context, uncond_context], axis=0)
+  noise_shape = (1, 4, height // 8, width // 8)
+
+  # Initialization starts from input_image if any, otherwise, starts from a random sampling.
+  if input_image:
+    if not hasattr(model, 'encoder'):
+      raise AttributeError(
+          'Stable Diffusion must be initilaized with encoder to accept input_image.'
+      )
+    input_image = input_image.resize((width, height))
+    input_image_np = np.array(input_image).astype(np.float32)
+    input_image_np = util.rescale(input_image, (0, 255), (-1, 1))
+    input_image_np = util.move_channel(input_image_np, to='first')
+    encoder_noise = np.random.normal(size=noise_shape).astype(np.float32)
+    latents = model.encoder(input_image_np, encoder_noise)
+    latents_noise = np.random.normal(size=noise_shape).astype(np.float32)
+    sampler.set_strength(strength=strength)
+    latents += latents_noise * sampler.initial_scale
+  else:
+    latents = np.random.normal(size=noise_shape).astype(np.float32)
+    latents *= sampler.initial_scale
+
+  # Diffusion process.
+  timesteps = tqdm(sampler.timesteps)
+  for i, timestep in enumerate(timesteps):
+    time_embedding = util.get_time_embedding(timestep)
+
+    input_latents = latents * sampler.get_input_scale()
+    input_latents = input_latents.repeat(2, axis=0)
+    output = model.diffusion(
+        input_latents, context, time_embedding, signature_name='diffusion'
+    )
+    output_cond, output_uncond = np.split(output, 2, axis=0)
+    output = cfg_scale * (output_cond - output_uncond) + output_uncond
+
+    latents = sampler.step(latents, output)
+
+  # Image decoding.
+  images = model.decoder(latents, signature_name='decode')
+  images = util.rescale(images, (-1, 1), (0, 255), clamp=True)
+  images = util.move_channel(images, to='last')
+  if not os.path.exists(output_path):
+    Path(output_path).parent.mkdir(parents=True, exist_ok=True)
+  Image.fromarray(images[0].astype(np.uint8)).save(output_path)
+
+
+if __name__ == '__main__':
+  run_tflite_pipeline(
+      StableDiffusion(
+          tokenizer_vocab_dir=os.path.join(
+              Path.home(), 'Downloads/stable_diffusion_data'
```

Comment:
can you allow user to provide a custom path ?

Reply:
Done. Update the script arguments.